### PR TITLE
fix: handle file compression check and test

### PR DESCRIPTION
**Pull request by jajjibhai008**
**Description**
Add file compression option on Reporting Form of frontend-admin-panel. Initially, that option was only enabled by the Django admin, not from the user request that came from the admin panel. But now users will be able to Enable compression from UI.

I apply validations on enable_compression by making a static function of compression_validation in models and using it in the serializer to make it secure.

Here is the picture of adding a button on the UI.

image

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/api/v1/serializers.py [hunk 1]
```diff
@@ -779,6 +780,7 @@ class Meta:
         default=[],
         child=serializers.EmailField()
     )
+    enable_compression = serializers.BooleanField()
```
Comment (saleem-latif) [inline enterprise/api/v1/serializers.py:783]: Why do we need to add `enable_compression` here? wouldn't this be added automatically by DRF?

File: enterprise/models.py [hunk 2]
```diff
@@ -2733,13 +2733,38 @@ def __repr__(self):
...
+        Args:
+            enable_compression (bool): file copression flag
+            data_type (string): report type
+            delivery_method (string): delivery method for sending files
+
+        Returns:
+           Validation Error
+        """
+
+        if not enable_compression and (
```
Comment (saleem-latif) [inline enterprise/models.py:2749]: can we simplify the condition here? I think following will be more readable.
```
if data_type in cls.ALLOWED_NON_COMPRESSION_DATA_TYPES or delivery_method == cls.DELIVERY_METHOD_SFTP:
    # Compression not needed and the can be disabled by the user.
    return
elif not enable_compression:
    allowed_data_types = ", ".join(cls.ALLOWED_NON_COMPRESSION_DATA_TYPES)
    error_message = (
        f'Compression can only be disabled for the following data types: {allowed_data_types} and '
         f'delivery method: {cls.DELIVERY_METHOD_SFTP}'
    )
    return {'enable_compression': error_message}
```

File: enterprise/models.py [hunk 3]
```diff
@@ -2733,13 +2733,38 @@ def __repr__(self):
         """
         return self.__str__()
 
+    @classmethod
+    def validate_compression(cls, enable_compression, data_type, delivery_method):
+        """Check enable_compression flag is set as expected
```
Comment (saleem-latif) [inline enterprise/models.py:2738]: Please update doc string format according to https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#docstrings

File: enterprise/models.py [hunk 4]
```diff
@@ -2733,13 +2733,38 @@ def __repr__(self):
...
+        return {}
+
     def clean(self):
         """
         Override of clean method to perform additional validation on frequency, day_of_month/day_of week
         and compression.
         """
         validation_errors = {}
-
+        print('enable compression ki value', self.enable_compression)
```
Comment (saleem-latif) [inline enterprise/models.py:2767]: nit: Please remove print statement.

**Review by saleem-latif — CHANGES_REQUESTED** [review]

File: enterprise/api/v1/serializers.py (hunk 1 above)
Comment (jajjibhai008) [reply enterprise/api/v1/serializers.py]: yes, you are right. I will update this.

File: enterprise/models.py (hunk 2 above)
Comment (jajjibhai008) [reply enterprise/models.py]: This looks good to me but there needs some changes in logic.
```
if data_type in cls.ALLOWED_NON_COMPRESSION_DATA_TYPES and delivery_method == cls.DELIVERY_METHOD_SFTP:
    # Compression not needed and the can be disabled by the user.
    return {}
elif not enable_compression:
    allowed_data_types = ", ".join(cls.ALLOWED_NON_COMPRESSION_DATA_TYPES)
    error_message = (
        f'Compression can only be disabled for the following data types: {allowed_data_types} and '
         f'delivery method: {cls.DELIVERY_METHOD_SFTP}'
    )
    return {'enable_compression': error_message}
return {}
```

**Review by saleem-latif — APPROVED** [review]

File: tests/test_enterprise/api/test_serializers.py [hunk 5]
```diff
@@ -337,3 +338,8 @@ def test_create(self):
         self.data['pgp_encryption_key'] = 'invalid-key'
         serializer = EnterpriseCustomerReportingConfigurationSerializer(data=self.data)
         assert not serializer.is_valid()
+
+        # Valid Compression check should be flagged.
+        self.data['enable_compression'] = False
+        serializer = EnterpriseCustomerReportingConfigurationSerializer(data=self.data)
+        assert not serializer.is_valid()
```
Comment (HammadAhmadWaqas) [inline tests/test_enterprise/api/test_serializers.py:345]: Check error message here, It should have expected error message.

File: enterprise/models.py [hunk 6]
```diff
@@ -2733,13 +2733,36 @@ def __repr__(self):
         """
         return self.__str__()
 
+    @classmethod
+    def validate_compression(cls, enable_compression, data_type, delivery_method):
```
Comment (HammadAhmadWaqas) [inline enterprise/models.py:2737]: Add a separate test for this method and pass different scenarios using `ddt`.

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: enterprise/models.py (hunk 6 above)
Comment (jajjibhai008) [reply enterprise/models.py:2737]: Already a test case exists in test_models.py for this method.

File: tests/test_enterprise/api/test_serializers.py (hunk 5 above)
Comment (jajjibhai008) [reply tests/test_enterprise/api/test_serializers.py:346]: Good idea to make it more efficient.

**Review by HammadAhmadWaqas — APPROVED** [review]

**Pull request merged**